velox/functions/prestosql/ArrayRemove.cpp

```diff
@@ -0,0 +1,288 @@
+    DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    vector_size_t& indicesCursor,
+    BufferPtr& newIndices,
+    BufferPtr& newElementNulls,
+    BufferPtr& newLengths,
+    BufferPtr& newOffsets) {
+  applyComplexType(
+      rows,
+      arrayDecoded,
+      elementsDecoded,
+      searchDecoded,
+      indicesCursor,
+      newIndices,
+      newElementNulls,
+      newLengths,
+      newOffsets);
+}
+
+template <>
+void applyTyped<TypeKind::ROW>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    vector_size_t& indicesCursor,
+    BufferPtr& newIndices,
+    BufferPtr& newElementNulls,
+    BufferPtr& newLengths,
+    BufferPtr& newOffsets) {
+  applyComplexType(
+      rows,
+      arrayDecoded,
+      elementsDecoded,
+      searchDecoded,
+      indicesCursor,
+      newIndices,
+      newElementNulls,
+      newLengths,
+      newOffsets);
+}
+
+class ArrayRemoveFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_USER_CHECK_EQ(
+        args.size(), 2, "array_remove requires exactly 2 parameters");
+
+    const auto& arrayVector = args[0];
+    const auto& searchVector = args[1];
+
+    // Both vectors must be array vectors of the same type.
+    VELOX_USER_CHECK(
+        arrayVector->type()->isArray(),
+        "array_remove requires arguments of type ARRAY");
+    VELOX_CHECK(
+        arrayVector->type()->asArray().elementType()->kindEquals(
+            searchVector->type()),
+        "array_remove requires all arguments of the same type: {} vs. {}",
+        searchVector->type(),
+        arrayVector->type()->asArray().elementType());
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+
+    DecodedVector* arrayDecoded = decodedArgs.at(0);
+    auto base = arrayDecoded->base()->as<ArrayVector>();
+
+    exec::LocalSelectivityVector nestedRows(context, base->elements()->size());
+    nestedRows->setAll();
+    exec::LocalDecodedVector elementsDecoded(
+        context, *base->elements(), *nestedRows);
+    DecodedVector* searchDecoded = decodedArgs.at(1);
+
+    vector_size_t elementsCount =
+        countElements<ArrayVector>(rows, *arrayDecoded);
+    vector_size_t rowCount = arrayDecoded->size();
+
+    // Allocate new vectors for indices, length, and offsets.
+    memory::MemoryPool* pool = context.pool();
+    BufferPtr newIndices = allocateIndices(elementsCount, pool);
+    BufferPtr newElementNulls =
+        AlignedBuffer::allocate<bool>(elementsCount, pool, bits::kNotNull);
+    BufferPtr newLengths = allocateSizes(rowCount, pool);
+    BufferPtr newOffsets = allocateOffsets(rowCount, pool);
+
+    vector_size_t indicesCursor = 0;
+
+    VELOX_DYNAMIC_TYPE_DISPATCH(
```

Comment:
The following functions use this method in a similar context:
- ArrayContains.cpp
- ArrayDistinct.cpp uses VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH
- ArrayDuplicates.cpp uses VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH

Is there a concern with using this macro? What is the alternative of having the function for different types?